research/minigo/dual_net.py

```diff
@@ -0,0 +1,209 @@
+    weights that match the graph. Used to set the weights to a different version
+    of the player without redefining the entire graph.
+
+    Args:
+      save_file: Path where model parameters were previously saved.
+    """
+
+    tf.train.Saver().restore(self.sess, save_file)
+
+  def run(self, position, use_random_symmetry=True):
+    """Compute the policy and value output for a given position."""
+    probs, values = self.run_many(
+        [position], use_random_symmetry=use_random_symmetry)
+    return probs[0], values[0]
+
+  def run_many(self, positions, use_random_symmetry=True):
+    processed = list(map(features.extract_features, positions))
+    if use_random_symmetry:
+      syms_used, processed = symmetries.randomize_symmetries_feat(processed)
+    outputs = self.sess.run(
+        self.inference_output, feed_dict={self.inference_input: processed})
+    probabilities, value = outputs['policy_output'], outputs['value_output']
+    if use_random_symmetry:
+      probabilities = symmetries.invert_symmetries_pi(syms_used, probabilities)
+    return probabilities, value
+
+
+def get_inference_input():
+  """Set up placeholders for input features/labels.
+
+  Returns:
+    The features and output tensors that get passed into model_fn.
+  """
+  input_features = tf.placeholder(
+      tf.float32, [None, go.N, go.N, features.NEW_FEATURES_PLANES],
+      name='pos_tensor')
+
+  labels = {
+      'pi_tensor': tf.placeholder(tf.float32, [None, go.N * go.N + 1]),
+      'value_tensor': tf.placeholder(tf.float32, [None])
+  }
+
+  return input_features, labels
+
+
+def bootstrap(working_dir, params):
+  """Initialize a tf.Estimator run with random initial weights.
+
+  Args:
+    working_dir: The directory where tf.estimator will drop logs,
+      checkpoints, and so on
+    params: hyperparams of the model.
+  """
+  # Forge an initial checkpoint with the name that subsequent Estimator will
+  # expect to find.
+  estimator_initial_checkpoint_name = 'model.ckpt-1'
+  save_file = os.path.join(working_dir,
+                           estimator_initial_checkpoint_name)
+  sess = tf.Session(graph=tf.Graph())
+  with sess.graph.as_default():
+    input_features, labels = get_inference_input()
+    dualnet_model.model_fn(
+        input_features, labels, tf.estimator.ModeKeys.PREDICT, params)
+    sess.run(tf.global_variables_initializer())
+    tf.train.Saver().save(sess, save_file)
+
+
+def export_model(working_dir, model_path):
+  """Take the latest checkpoint and export it to model_path for selfplay.
+
+  Assumes that all relevant model files are prefixed by the same name.
+  (For example, foo.index, foo.meta and foo.data-00000-of-00001).
+
+  Args:
+    working_dir: The directory where tf.estimator keeps its checkpoints.
+    model_path: Either a local path or a gs:// path to export model to.
+  """
+  estimator = tf.estimator.Estimator(
+      dualnet_model.model_fn, model_dir=working_dir, params='ignored')
```

Comment:
You should be able to avoid creating a fake estimator-- tf.train.latest_checkpoint(working_dir) is the same.